lib/WUI/espif.c

```diff
@@ -0,0 +1,521 @@
+        return 0;
+    }
+}
+
+static err_t espif_reconfigure_uart(const uint32_t baudrate) {
+    log_info(ESPIF, "Reconfiguring UART for %d baud", baudrate);
+    huart6.Init.BaudRate = baudrate;
+    int hal_uart_res = HAL_UART_Init(&huart6);
+    if (hal_uart_res != HAL_OK) {
+        log_error(ESPIF, "ESP LL: HAL_UART_Init failed with: %d", hal_uart_res);
+        return ERR_IF;
+    }
+
+    int hal_dma_res = HAL_UART_Receive_DMA(&huart6, (uint8_t *)dma_buffer_rx, RX_BUFFER_LEN);
+    if (hal_dma_res != HAL_OK) {
+        log_error(ESPIF, "ESP LL: HAL_UART_Receive_DMA failed with: %d", hal_dma_res);
+        return ERR_IF;
+    }
+
+    return ERR_OK;
+}
+
+static err_t low_level_init(struct netif *netif) {
+    if (!initialized) {
+        espif_reconfigure_uart(ESPIF_UART_BAUDRATE);
+
+        // Create mbox and start thread
+        if (uartBufferMbox_id == NULL) {
+            uartBufferMbox_id = osMessageCreate(osMessageQ(uartBufferMbox), NULL);
+            if (uartBufferMbox_id == NULL) {
+                log_error(ESPIF, "ESP LL: failed to create UART buffer mbox");
+                return ERR_MEM;
+            }
+        }
+        if (UartBufferThread_id == NULL) {
+            UartBufferThread_id = osThreadCreate(osThread(UartBufferThread), netif);
+            if (UartBufferThread_id == NULL) {
+                log_error(ESPIF, "error!");
+                return ERR_MEM;
+            }
+        }
+    } else {
+        log_error(ESPIF, "Already initialized");
+    }
+
+    espif_set_operating_mode(ESPIF_RUNNING_MODE);
+    initialized = 1;
+
+    return ERR_OK;
+}
+
+static err_t low_level_deinit() {
+    if (uartBufferMbox_id != NULL) {
+        osMessageQId tmp = uartBufferMbox_id;
+        uartBufferMbox_id = NULL;
+        osMessageDelete(tmp);
+    }
+    if (UartBufferThread_id != NULL) {
+        osThreadId tmp = UartBufferThread_id;
+        UartBufferThread_id = NULL;
+        osThreadTerminate(tmp);
+    }
+    initialized = 0;
+    return ERR_OK;
+}
+
+err_t espif_flash_initialize() {
+    err_t err = low_level_deinit();
+    if (err != ERR_OK) {
+        return err;
+    }
+    espif_set_operating_mode(ESPIF_FLASHING_MODE);
+    espif_reconfigure_uart(115200);
+    loader_stm32_config_t loader_config = {
+        .huart = &huart6,
+        .port_io0 = GPIOE,
+        .pin_num_io0 = GPIO_PIN_6,
+        .port_rst = GPIOC,
+        .pin_num_rst = GPIO_PIN_13,
+    };
+    loader_port_stm32_init(&loader_config);
+    return ERR_OK;
+}
+
+/**
+ * \brief           USART data processing
+ */
+static void uart_reader(void const *arg) {
+    struct netif *netif = (struct netif *)arg;
+
+    size_t pos = 0;
+
+    log_info(ESPIF, "Reader thread running");
+
+    while (1) {
+        /* Wait for the event message from DMA or USART */
+        /* There is 100ms max wait time to ensure some small standalone message
+           does not get stuck in DMA buffer for too long. */
+        osMessageGet(uartBufferMbox_id, 100);
```

Comment:
Currently it is kept running. With message box replaced by signal/wait it turned out complex to handle destroying thread that can be woken up by ISR.
There is a condition on operation mode (not a strict synchronization like mutex) embedded in position checking so that this thread does nothing when flashing. In theory it can be already reading chunk of input so that is can move pointer position afterwards, but that would mean it is reading legitimate data received before entering the flash mode.

There is an option to introduce "uart_reader_mutex" (there already is a "uart_writer_mutex"). This one would be held exclusively by reader thread during normal operation while during flashing both reader and writer mutexes would be taken by flasher. The problem is that flasher is an external entity here, we only receive calls for enter/leave flashing mode. THis can be worked out somehow, of course. Sendond point against this is that this would bring overhead of locking at almost no benefit. When the ESP is reset a pile of garbage is going to end up in the UART anyway,